nullaway/src/main/java/com/uber/nullaway/dataflow/AccessPath.java

```diff
@@ -169,11 +172,44 @@ public static AccessPath getForMapInvocation(MethodInvocationNode node) {
     return fromMapGetCall(node);
   }
 
+  @Nullable
+  private static MapKey argumentToMapKeySpecifier(Node argument) {
+    // A switch at the Tree level should be faster than multiple if checks at the Node level.
+    switch (argument.getTree().getKind()) {
+      case STRING_LITERAL:
+        return new StringMapKey(((StringLiteralNode) argument).getValue());
+      case INT_LITERAL:
+        return new NumericMapKey(((IntegerLiteralNode) argument).getValue());
+      case LONG_LITERAL:
+        return new NumericMapKey(((LongLiteralNode) argument).getValue());
+      case METHOD_INVOCATION:
+        MethodAccessNode target = ((MethodInvocationNode) argument).getTarget();
+        List<Node> arguments = ((MethodInvocationNode) argument).getArguments();
+        // Check for int/long boxing.
+        if (target.getMethod().getSimpleName().toString().equals("valueOf")
+            && arguments.size() == 1
+            && target.getReceiver().getTree().getKind().equals(Tree.Kind.IDENTIFIER)
+            && (target.getReceiver().toString().equals("Integer")
+                || target.getReceiver().toString().equals("Long"))) {
+          return argumentToMapKeySpecifier(arguments.get(0));
+        }
+        // Fine to fallthrough:
+      default:
+        // Every other type of expression, including variables, field accesses, new A(...), etc.
+        AccessPath accessPath = getAccessPathForNodeNoMapGet(argument);
+        if (accessPath == null) {
+          // No MapKey object can be constructed here
+          return null;
+        }
+        return new AccessPathMapKey(accessPath);
```

Comment:
Do you think it would be worthwhile to just have `AccessPath` implement `MapKey` instead of introducing this `AccessPathMapKey` wrapper?  I don't know what the common case will be here or what the perf impact would be, but that would save one level of indirection for `AccessPath`s.

Reply:
Sounds good to me. It reduces runtime indirection _and_ code size :)